v0/destinations/refiner/utils.js

```diff
@@ -0,0 +1,259 @@
+    if (payload[key]) {
+      const value = payload[key];
+      payload[attributesMap[key]] = value;
+      delete payload[key];
+    }
+  });
+  return payload;
+};
+
+/**
+ * Returns account object
+ * ref : https://refiner.io/docs/api/#group-users
+ * @param {*} message
+ * @param {*} destination
+ * @returns
+ */
+const getAccountTraits = (message, destination) => {
+  const { accountAttributesMapping } = destination.Config;
+  // all the traits fields, that are not mapped in webapp are going as it is
+  const traits = getFieldValueFromMessage(message, "groupTraits");
+  const id = getFieldValueFromMessage(message, "groupId");
+  let groupUsersPayload = {
+    ...traits,
+    id
+  };
+  const accountAttributesMap = getHashFromArray(
+    accountAttributesMapping,
+    "from",
+    "to",
+    false
+  );
+  groupUsersPayload = replaceDestAttributes(
+    accountAttributesMap,
+    groupUsersPayload
+  );
+  return groupUsersPayload;
+};
+
+/**
+ * Returns identify event payload
+ * ref : https://refiner.io/docs/api/#identify-user
+ * @param {*} message
+ * @param {*} destination
+ * @returns
+ */
+const identifyUserPayloadBuilder = (message, destination) => {
+  const { userAttributesMapping } = destination.Config;
+  // mapping refiner remote_id with userId
+  const userId = getFieldValueFromMessage(message, "userIdOnly");
+  const email = getFieldValueFromMessage(message, "email");
+  // all the traits fields, that are not mapped in webapp are going as it is
+  const traits = getFieldValueFromMessage(message, "traits");
+  const contextTraits = get(message, "context.traits");
+  let payload = {
+    ...traits,
+    ...contextTraits,
+    userId,
+    email
+  };
+  const userAttributesMap = getHashFromArray(
+    userAttributesMapping,
+    "from",
+    "to",
+    false
+  );
+  payload = replaceDestAttributes(userAttributesMap, payload);
+  const { endpoint } = CONFIG_CATEGORIES.IDENTIFY_USER;
+  return {
+    payload,
+    endpoint
+  };
+};
+
+/**
+ * Returns track event payload
+ * ref : https://refiner.io/docs/api/#track-event
+ * @param {*} message
+ * @returns
+ */
+const trackEventPayloadBuilder = message => {
+  const payload = {};
+  const userId = getFieldValueFromMessage(message, "userIdOnly");
+  const email = getFieldValueFromMessage(message, "email");
+  const { event } = message;
+  if (userId) {
+    payload.id = userId;
+  }
+  if (email) {
+    payload.email = email;
+  }
+  payload.event = event;
+  const { endpoint } = CONFIG_CATEGORIES.TRACK_EVENT;
+  return { payload, endpoint };
+};
+
+/**
+ * Returns page event payload
+ * ref : https://refiner.io/docs/api/#track-event
+ * @param {*} message
+ * @returns
+ */
+const pageEventPayloadBuilder = message => {
+  const payload = {};
+  let event;
+  const userId = getFieldValueFromMessage(message, "userIdOnly");
+  const email = getFieldValueFromMessage(message, "email");
+  if (userId) {
+    payload.id = userId;
+  }
+  if (email) {
+    payload.email = email;
+  }
```

Comment:
same here, while any of id or email is not present, if we cannot send event, we need to throw error

Reply:
@shrouti1507 , if you go to parent function of this function which is **pageResponseBuilder**, there we are calling a function called   **validatePayload** which is responsible for validating the payload and throwing an error. so if we are coming here then it means it has either one is present otherwise we already thrown an error and same is applicable for all the calls

Reply:
got it.